### PR TITLE
Port symbol.js to @flow.

### LINKED ISSUE #857: Port symbols.js to @flow

### PR DESCRIPTION
**Pull request by marcianx**
Fixed #857.

### CODE REVIEW

**Comment by marcianx** [comment]
Actually, wait a sec. The `mode` is used in `ParserNode` as well, so I'll pull it out into a separate file that just defines the `mode` type.

**Comment by marcianx** [comment]
Pulled out  `Mode`. I wasn't sure whether I should put it in `types.js` or `Mode.js`. Types cannot be default exported, so `import {Mode} from "./Mode";` didn't seem as nice. But that's what I've gone with.

**Comment by marcianx** [comment]
Changed my mind again and renamed `Mode.js` to `types.js` since I foresee adding a `ParserInterface` type to break circular dependencies.

File: src/symbols.js [hunk 1]
```diff
@@ -16,20 +17,28 @@
...
-function defineSymbol(mode, font, group, replace, name, acceptUnicodeChar) {
-    module.exports[mode][name] = {
-        font: font,
-        group: group,
-        replace: replace,
-    };
+/** `acceptUnicodeChar = true` is only applicable if `replace` is set. */
+function defineSymbol(
+    mode: Mode, font: Font, group: Group, replace: ?string, name: string,
+    acceptUnicodeChar?: boolean) {
```
Comment (kevinbarabash) [inline src/symbols.js]: Ideally I'd like function args to formatted as so it's easier to read the args:
```
function foo(
   arg0: T0,
   arg1: T1,
   ...
   argn: Tn,
) {
   console.log('bar');
   ...
}
```
Unfortunately, I don't believe our babel setup supports trailing commas in function defintions and calls so let's leave this for now.  I opened an issue for this: https://github.com/Khan/KaTeX/issues/861.

File: src/types.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// @flow
+
+/**
+ * This file consists only of basic flow types used in multiple places.
+ * For types with javascript, create separate files by themselves.
```
Comment (kevinbarabash) [inline src/types.js]: Nice!

File: src/ParseNode.js [hunk 3]
```diff
@@ -23,7 +24,7 @@ export default class ParseNode {
     constructor(
         type: string,       // type of node, like e.g. "ordgroup"
         value: mixed,       // type-specific representation of the node
```
Comment (kevinbarabash) [inline src/ParseNode.js]: I just noticed this was `mixed`.  Hopefully we'll be able to make this more precise in the future.

Comment (marcianx) [reply src/ParseNode.js]: I agree. A lot of refactoring needs to be done before we can do that.

**Pull request merged**